### PR TITLE
planner: consider prefix index column length in skyline pruning

### LINKED ISSUE #27313: Skyline pruning prunes maybe-good prefix index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
use test;
drop table if exists t;
create table t(a varchar(100), b int, c int, index idx1(a(2), b), index idx2(a));
explain format = 'verbose' select * from t where a = 'abcdefghijk' and b > 4;
```

### 2. What did you expect to see? (Required)
`idx2` remains after skyline pruning.

### 3. What did you see instead (Required)
`idx2` is pruned since skyline pruning thinks it is worse than `idx1`, which is not true considering the index column prefix length.
```
mysql> explain format = 'verbose' select * from t where a = 'abcdefghijk' and b > 4;
+-------------------------------+---------+---------+-----------+---------------------------+----------------------------------------------------------+
| id                            | estRows | estCost | task      | access object             | operator info                                            |
+-------------------------------+---------+---------+-----------+---------------------------+----------------------------------------------------------+
| IndexLookUp_11                | 33.33   | 678.84  | root      |                           |                                                          |
| ├─IndexRangeScan_8(Build)     | 33.33   | 2370.00 | cop[tikv] | table:t, index:idx1(a, b) | range:("ab" 4,"ab" +inf], keep order:false, stats:pseudo |
| └─Selection_10(Probe)         | 33.33   | 0.00    | cop[tikv] |                           | eq(test.t.a, "abcdefghijk")                              |
|   └─TableRowIDScan_9          | 33.33   | 2370.00 | cop[tikv] | table:t                   | keep order:false, stats:pseudo                           |
+-------------------------------+---------+---------+-----------+---------------------------+----------------------------------------------------------+
4 rows in set, 1 warning (0.00 sec)

mysql> show warnings;
+-------+------+---------------------------------------------------------------------------------------+
| Level | Code | Message                                                                               |
+-------+------+---------------------------------------------------------------------------------------+
| Note  | 1105 | [t,idx1] remain after pruning paths for t given Prop{SortItems: [], TaskTp: rootTask} |
+-------+------+---------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)
```
mysql> SELECT tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                            |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.2.0-alpha-653-g4a3c290c0
Edition: Community
Git Commit Hash: 4a3c290c0b6f80d9d3eea430a59bfc2f9cb9c78f
Git Branch: master
UTC Build Time: 2021-08-17 11:10:12
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #27313 

Problem Summary:
Skyline pruning prunes maybe-good prefix index.

### What is changed and how it works?

When comparing columns occurred in access conditions or index conditions, we say `path1`'s columns are a strictly subset of  `path2`'s if each `path1`'s column exists in `path2`'s columns and the length of `path1`'s column is less than or equal to the length of the corresponding `path2`'s column.

### Check List 

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
'None'
```

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -415,12 +416,112 @@ func (ds *DataSource) tryToGetDualTask() (task, error) {
...
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+func compareColumnsWithLens(lhs, rhs []*columnWithLen) (int, bool) {
```
Comment (winoros) [inline planner/core/find_best_task.go:505]: Add some comments for this function.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -415,12 +416,112 @@ func (ds *DataSource) tryToGetDualTask() (task, error) {
 	return nil, nil
 }
 
+type columnWithLen struct {
```
Comment (winoros) [inline planner/core/find_best_task.go:419]: We can move it to `planner/util`.
And add type alias `type ColumnsWithLen []*ColumnWithLen`
Then move the functions `firstDominateSecond`, `compareColumnsWithLens` to `ColumnsWithLen`'s member function.

File: planner/core/find_best_task.go [hunk 3]
```diff
@@ -415,12 +416,112 @@ func (ds *DataSource) tryToGetDualTask() (task, error) {
...
+	if l < r {
+		return -1
+	}
+	if l == r {
+		return 0
+	}
+	return 1
+}
+
+func firstDominateSecond(cols1, cols2 []*columnWithLen) bool {
```
Comment (winoros) [inline planner/core/find_best_task.go:481]: Also, add comments for this one.

File: planner/util/path_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
```
Comment (winoros) [inline planner/util/path_test.go:19]: Split the import package.

File: planner/util/path.go [hunk 5]
```diff
@@ -181,3 +181,105 @@ func (path *AccessPath) OnlyPointRange(sctx sessionctx.Context) bool {
...
+	if l > r {
+		return 1
+	}
+	if l == r {
+		return 0
+	}
+	return -1
+}
+
+// c1 dominates c2 when each column of c2 exists in c1 and c2's column length is no longer than c1's column length.
```
Comment (winoros) [inline planner/util/path.go:237]: Maybe we should follow the pattern
```suggestion
// dominates checks/tells ...
```

**Review by winoros — APPROVED** [review]

File: planner/util/path.go [hunk 6]
```diff
@@ -181,3 +181,106 @@ func (path *AccessPath) OnlyPointRange(sctx sessionctx.Context) bool {
...
+			}
+		}
+	case *expression.ScalarFunction:
+		for _, arg := range v.GetArgs() {
+			extractCol2LenFromExpr(arg, idxCols, idxColLens, col2Len)
+		}
+	}
+}
+
+func compareLength(l, r int) int {
```
Comment (Reminiscent) [inline planner/util/path.go:218]: Add some comments to explain the return value.

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**